src/ess/reflectometry/tools.py

```diff
@@ -108,3 +112,160 @@ def linlogspace(
         grids.append(mesh[dim, start:])
 
     return sc.concat(grids, dim)
+
+
+def _sort_by(a, by):
+    return [x for x, _ in sorted(zip(a, by, strict=True), key=lambda x: x[1])]
+
+
+def _find_interval_overlaps(intervals):
+    '''Returns the intervals where at least
+    two or more of the provided intervals
+    are overlapping.'''
+    edges = list(chain.from_iterable(intervals))
+    is_start_edge = list(chain.from_iterable((True, False) for _ in intervals))
+    edges_sorted = sorted(edges)
+    is_start_edge_sorted = _sort_by(is_start_edge, edges)
+
+    number_overlapping = 0
+    overlap_intervals = []
+    for x, is_start in zip(edges_sorted, is_start_edge_sorted, strict=True):
+        if number_overlapping == 1 and is_start:
+            start = x
+        if number_overlapping == 2 and not is_start:
+            overlap_intervals.append((start, x))
+        if is_start:
+            number_overlapping += 1
+        else:
+            number_overlapping -= 1
+    return overlap_intervals
+
+
+def _searchsorted(a, v):
+    for i, e in enumerate(a):
+        if e > v:
+            return i
+    return len(a)
+
+
+def _create_qgrid_where_overlapping(qgrids):
+    '''Given a number of Q-grids, construct a new grid
+    covering the regions where (any two of the) provided grids overlap.'''
+    pieces = []
+    for start, end in _find_interval_overlaps([(q.min(), q.max()) for q in qgrids]):
+        interval_sliced_from_qgrids = [
+            q[max(_searchsorted(q, start) - 1, 0) : _searchsorted(q, end) + 1]
+            for q in qgrids
+        ]
+        densest_grid_in_interval = max(interval_sliced_from_qgrids, key=len)
+        pieces.append(densest_grid_in_interval)
+    return sc.concat(pieces, dim='Q')
+
+
+def _interpolate_on_qgrid(f, curves, grid):
+    return np.stack(
+        [
+            np.interp(
```

Comment:
Can we use https://scipp.github.io/generated/modules/scipy/scipp.scipy.interpolate.html ?

Reply:
Yeah probably, I can look into it.